units/sdlthread.inc

```diff
@@ -111,123 +145,261 @@ function SDL_CreateThread(fn: TSDL_ThreadFunction; name: PAnsiChar;
-   *
-   *  var tls_lock: TSDL_SpinLock;
-   *    thread_local_storage: TSDL_TLSID;
-   *
-   *  procedure SetMyThreadData(value: Pointer)
-   *
-   *      if not (thread_local_storage) then
-   *      begin
-   *        SDL_AtomicLock(@tls_lock);
-   *        if (!thread_local_storage)
-   *              thread_local_storage = SDL_TLSCreate();
-   *
-   *        SDL_AtomicUnLock(@tls_lock);
-   *
-   *        SDL_TLSSet(thread_local_storage, value);
-   *      end;
-   *
-   *  function GetMyThreadData(): Pointer;
-   *  begin
-   *     Result := SDL_TLSGet(thread_local_storage);
-   *  end;
-   *
-   *   SDL_TLSGet()
-   *   SDL_TLSSet()
-   *}
-function SDL_TLSCreate: TSDL_TLSID cdecl; external SDL_LibName {$IFDEF DELPHI} {$IFDEF MACOS} name '_SDL_TLSCreate' {$ENDIF} {$ENDIF};
+ * Create a piece of thread-local storage.
+ *
+ * This creates an identifier that is globally visible to all threads but
+ * refers to data that is thread-specific.
+ *
+ * \returns the newly created thread local storage identifier or 0 on error.
+ *
+ * \since This function is available since SDL 2.0.0.
+ *
+ * \sa SDL_TLSGet
+ * \sa SDL_TLSSet
+ *}
+function SDL_TLSCreate: TSDL_TLSID; cdecl;
+  external SDL_LibName {$IFDEF DELPHI} {$IFDEF MACOS} name '_SDL_TLSCreate' {$ENDIF} {$ENDIF};
 
   {**
-   *  Get the value associated with a thread local storage ID for the current thread.
-   *
-   *   id The thread local storage ID
-   *
-   *   The value associated with the ID for the current thread, or NULL if no value has been set.
-   *
-   *   SDL_TLSCreate()
-   *   SDL_TLSSet()
-   *}
-function SDL_TLSGet(id: TSDL_TLSID): Pointer cdecl; external SDL_LibName {$IFDEF DELPHI} {$IFDEF MACOS} name '_SDL_TLSGet' {$ENDIF} {$ENDIF};
+ * Get the current thread's value associated with a thread local storage ID.
+ *
+ * \param id the thread local storage ID
+ * \returns the value associated with the ID for the current thread or NULL if
+ *          no value has been set; call SDL_GetError() for more information.
+ *
+ * \since This function is available since SDL 2.0.0.
+ *
+ * \sa SDL_TLSCreate
+ * \sa SDL_TLSSet
+ *}
+function SDL_TLSGet(id: TSDL_TLSID): Pointer; cdecl;
+  external SDL_LibName {$IFDEF DELPHI} {$IFDEF MACOS} name '_SDL_TLSGet' {$ENDIF} {$ENDIF};
 
   {**
-   *  Set the value associated with a thread local storage ID for the current thread.
-   *
-   *   id The thread local storage ID
-   *   value The value to associate with the ID for the current thread
-   *   destructor_ A function called when the thread exits, to free the value.
-   *
-   *   0 on success, -1 on error
-   *
-   *   SDL_TLSCreate()
-   *   SDL_TLSGet()
-   *}
-function SDL_TLSSet(id: TSDL_TLSID; value: Pointer; destructor_: Pointer): cint32 cdecl; external SDL_LibName {$IFDEF DELPHI} {$IFDEF MACOS} name '_SDL_TLSSet' {$ENDIF} {$ENDIF};
+ * Set the current thread's value associated with a thread local storage ID.
+ *
+ * The function prototype for `destructor` is:
```

Comment:
Something along the lines as following?

`TTLSDestructor = procedure(value: Pointer); cdecl;`

`function SDL_TLSSet(id: TSDL_TLSID; const value: Pointer; destructor_: TTLSDestructor): cint; cdecl;
  external SDL_LibName {$IFDEF DELPHI} {$IFDEF MACOS} name '_SDL_TLSSet' {$ENDIF} {$ENDIF};`

Would accompany this by a comment to explain why we did it this way.

Reply:
Yes, exactly.

Reply:
Thanks for your replies :). I have not a lot of time, so I reply only to this part at the moment.

On reconsidering this, I fear I do not like the idea. I fully agree that it would be beneficial to have type-checking, but 1) we generally try to stay as close as possible to the original C header code, 2) someone using threads should be able to make sure it is really a destructor function pointer by himself, 3) this would raise the question about other function pointers, how should we treat them, always introduce a specific type (or is there a special reason why we should introduce a type-check explicitly for this function pointer)?

Best regards
Matthias

Reply:
On closer examination I saw that the original headers use a function pointer (in contrast to a simple pointer), too. Therefore I implemented the function pointer as discussed here.